bundles/org.openhab.binding.heliosventilation/src/main/java/org/openhab/binding/heliosventilation/internal/HeliosVentilationHandler.java

```diff
@@ -0,0 +1,416 @@
+        IOUtils.closeQuietly(outputStream);
+        outputStream = null;
+        inputStream = null;
+        serialPort = null;
+    }
+
+    private void poll(HeliosVentilationDataPoint v) {
+        byte txFrame[] = { 0x01, BUSMEMBER_ME, BUSMEMBER_MAINBOARD, 0x00, v.address(), 0x00 };
+        txFrame[5] = (byte) checksum(txFrame);
+
+        tx(txFrame);
+    }
+
+    /*
+     * transmit a frame
+     */
+    private void tx(byte[] txFrame) {
+        try {
+            OutputStream out = outputStream;
+            if (out != null) {
+                logger.trace("HeliosVentilation: Write to serial port: {}",
+                        String.format("%02x %02x %02x %02x", txFrame[1], txFrame[2], txFrame[3], txFrame[4]));
+
+                out.write(txFrame);
+                // after each frame we have to wait.
+                // 30 ms is taken from what we roughly see the original remote control is doing
+                Thread.sleep(30);
+            }
+        } catch (IOException e) {
+            // in case we cannot write the connection is somehow broken, let's officially disconnect
+            disconnect();
+            connect();
+        } catch (InterruptedException e) {
+            // ignore if we got interrupted
+        }
+    }
+
+    /**
+     * Check connection status
+     *
+     * @return true if currently connected
+     */
+    private boolean isConnected() {
+        return serialPort != null && inputStream != null && outputStream != null;
+    }
+
+    @Override
+    public void serialEvent(SerialPortEvent event) {
+        switch (event.getEventType()) {
+            case SerialPortEvent.DATA_AVAILABLE:
+                // we get here if data has been received
+                try {
+                    try {
+                        // wait a little to increase the likelihood that the complete frame is already buffered.
+                        // this improves the robustness for RS485/USB converters which sometimes duplicate bytes
+                        // otherwise
+                        Thread.sleep(5);
+                    } catch (InterruptedException e) {
+                        // ignore interruption
+                    }
+                    byte frame[] = { 0, 0, 0, 0, 0, 0 };
+                    InputStream in = inputStream;
+                    if (in != null) {
+                        do {
+                            int cnt = 0;
+                            int c;
+                            do {
+                                c = in.read();
+                                if (cnt > 0 || c == 0x01) {
+                                    frame[cnt] = (byte) c;
+                                    cnt++;
+                                }
+
+                                if (cnt < 6 && in.available() < 1) {
+                                    // frame not yet complete but no input available, let's wait a little to merge
+                                    // interrupted transmissions
```

Comment:
If another DATA_AVAILABLE event arrives while waiting wouldn't it interfere with parsing this frame? I'm not too familiar with how the serial binding works so I might be completely off base here.

Reply:
From what I observed during debugging it doesn't seem that another DATA_AVAILABLE event will be handled in another thread and therefore preempting this one here. But my experience with the serial binding is also limited, so if you see the need to involve an expert here, don't hesitate.